benchto-driver/src/main/java/io/trino/benchto/driver/execution/BenchmarkExecutionDriver.java

```
@@ -158,19 +173,80 @@ private BenchmarkExecutionResult failedBenchmarkResult(Benchmark benchmark, Exce
+                for (int queryIndex = 0; queryIndex < benchmark.getQueries().size(); queryIndex++) {
+                    int permutedQueryIndex = queryIndex;
+                    if (!reportStatus) {
+                        if (queryIndex % benchmark.getConcurrency() != threadNumber) {
+                            // for pre-warming we split queries among all threads instead
+                            // of each thread running all queries
+                            continue;
+                        }
+                        LOG.info("Executing pre-warm query {}", queryIndex);
+                    }
+                    else {
+                        permutedQueryIndex = queryOrder[queryIndex];
+                    }
+                    Query query = benchmark.getQueries().get(permutedQueryIndex);
+                    if (isTimeLimitExceeded(executionTimeLimit)) {
+                        LOG.warn("Interrupting benchmark {} due to time limit exceeded", benchmark.getName());
+                        return queryExecutionResults;
+                    }
+                    QueryExecution queryExecution = new QueryExecution(benchmark, query,
+                            queryIndex
+                                    + threadNumber * benchmark.getQueries().size()
+                                    + (run - 1) * benchmark.getConcurrency() * benchmark.getQueries().size());
+                    if (firstQuery && reportStatus) {
+                        statusReporter.reportExecutionStarted(queryExecution);
+                        firstQuery = false;
+                    }
+                    Optional<QueryExecutionResult> result = executeSingleQuery(queryExecution, benchmark, connection, false, executionTimeLimit);
+                    result.ifPresent(queryExecutionResults::add);
+                }
+            }
+        }
+        return queryExecutionResults;
+    }
+
+    private Optional<QueryExecutionResult> executeSingleQuery(
```

Just throw some `TimeLimitException extends RuntimeException` instead of returning `Optional`.
`Optional` is kind of meaningless (was it OK or error?, what caused error?).